app/models/plan.rb

```diff
@@ -73,8 +73,6 @@ class PeriodRangeCalculationError < StandardError; end
   # But calling `#lock!` will call `#reload` so some instance variables are reset
   before_destroy :avoid_destruction, prepend: true
 
-  has_many :cinstances, :dependent => :destroy
```

Comment:
this association is redundant as it is better defined in the `ApplicaitonPlan` class. In general cinstances are associated to application plans. Contracts are the parent class for all kinds of contracts.

Reply:
I think it shouldn't be used but some tests fail. I suspect some wrong `inverse_of` relation though. Investigating.

Reply:
Actually not exactly. I saw where it is used 
```
self.background_deletion = [:cinstances, ...]
```

Reply:
There are 3 contract STI types:
```rb
[22] pry(main)> Contract.descendants.map(&:name)
=> ["Cinstance", "ServiceContract", "AccountContract"]
```

They correspond to the 3 plan types:
```rb
[23] pry(main)> Plan.descendants.map(&:name)
=> ["AccountPlan", "ServicePlan", "ApplicationPlan"]
```

So I've tried the result in rails console with the original code (`has_many :cinstances, :dependent => :destroy`)
```rb
[49] pry(main)> service_plan.contracts.reload ; nil
  ServiceContract Load (1.1ms)  SELECT `cinstances`.* FROM `cinstances` WHERE `cinstances`.`type` = 'ServiceContract' AND `cinstances`.`plan_id` = 1
[50] pry(main)> service_plan.cinstances.reload ; nil
  Cinstance Load (1.2ms)  SELECT `cinstances`.* FROM `cinstances` WHERE `cinstances`.`type` = 'Cinstance' AND `cinstances`.`plan_id` = 1

[51] pry(main)> account_plan.contracts.reload ; nil
  AccountContract Load (0.9ms)  SELECT `cinstances`.* FROM `cinstances` WHERE `cinstances`.`type` = 'AccountContract' AND `cinstances`.`plan_id` = 3
[52] pry(main)> account_plan.cinstances.reload ; nil
  Cinstance Load (0.9ms)  SELECT `cinstances`.* FROM `cinstances` WHERE `cinstances`.`type` = 'Cinstance' AND `cinstances`.`plan_id` = 3

[53] pry(main)> application_plan.contracts.reload ; nil
  Cinstance Load (1.0ms)  SELECT `cinstances`.* FROM `cinstances` WHERE `cinstances`.`type` = 'Cinstance' AND `cinstances`.`plan_id` = 2
[54] pry(main)> application_plan.cinstances.reload ; nil
  Cinstance Load (0.9ms)  SELECT `cinstances`.* FROM `cinstances` WHERE `cinstances`.`type` = 'Cinstance' AND `cinstances`.`plan_id` = 2
```

In all cases the `contracts` association is valid but the `cinstances` association is valid only for `ApplicationPlan`. That's why it is redundant to keep `:cinstances` in the `background_deletion` list. One cannot assign a `Cinstance` to any plan but an `ApplicationPlan`.

```rb
[55] pry(main)> FactoryBot.create(:cinstance, plan: service_plan)
ActiveRecord::AssociationTypeMismatch: ApplicationPlan(#34080) expected, got #<ServicePlan ...

[56] pry(main)> FactoryBot.create(:cinstance, plan: account_plan)
ActiveRecord::AssociationTypeMismatch: ApplicationPlan(#34080) expected, got #<AccountPlan ...
```

This means that `cinstances` will return empty in all but `ApplicationPlan` as well will return a needless duplicate of `contracts` for `ApplicationPlan`.

Furthermore any usage of `cinstances` elsewhere in the application on a non `ApplicationPlan` would return nothing and thus be invalid/already a bug. I'll go through the grep another time though to see if I spot anything.

Reply:
Case in point, explanations work when they are either smart or long.